File: server/internal/accounts/db/simplefin.go
package accountsdb

import (
	"context"
	"database/sql"
	"fmt"
	"time"

	"github.com/samber/lo"
	"tallyo/internal/accounts"
	"tallyo/internal/database/dbgen"
	"tallyo/internal/database/dbutil"
	"tallyo/internal/graph/model"
	u "tallyo/internal/utils"
)

// LinkSimpleFinConnection upserts the simplefin_connections row and its
// polymorphic connections row in one transaction, returning both IDs so
// callers (initial linking and every sync pass) never need to reread either.
func (s *Store) LinkSimpleFinConnection(ctx context.Context, conn accounts.UpsertSimpleFinConnectionParams) (simpleFinConnID int64, connectionID int64, err error) {
	err = s.WithTx(ctx, func(_ *sql.Tx, q *dbgen.Queries) error {
		id, err := q.UpsertSimpleFinConnection(ctx, dbgen.UpsertSimpleFinConnectionParams{
			ExternalID:    conn.ExternalID,
			AccessTokenID: conn.AccessTokenID,
			OrgID:         conn.OrgID,
			OrgDomain:     conn.OrgDomain,
			OrgUrl:        conn.OrgURL,
			SfinUrl:       conn.SfinURL,
			LogoUrl:       conn.LogoURL,
		})
		if err != nil {
			return err
		}
		simpleFinConnID = id
		connectionID, err = createSimpleFinConnectionRow(ctx, q, id, conn.Name, conn.OwnerID)
		return err
	})
	return simpleFinConnID, connectionID, err
}

func createSimpleFinConnectionRow(ctx context.Context, q *dbgen.Queries, connID int64, name string, ownerID int64) (int64, error) {
	id, err := q.UpsertConnection(ctx, dbgen.UpsertConnectionParams{
		SourceTable: string(accounts.SourceTableSimpleFinConnection),
		SourceID:    connID,
		Name:        lo.EmptyableToPtr(name),
		OwnerID:     ownerID,
	})
	if err != nil {
		return 0, fmt.Errorf("upsert simplefin connection row: %w", err)
	}
	return id, nil
}

// SimpleFinConnectionsByTokenIDs hydrates every token's connections (with
// their accounts) in a fixed two reads regardless of token count.
func (s *Store) SimpleFinConnectionsByTokenIDs(ctx context.Context, accessTokenIDs []int64) (map[int64][]*model.SimpleFinConnection, error) {
	toEmptyConnections := func(id int64) (int64, []*model.SimpleFinConnection) {
		return id, []*model.SimpleFinConnection{}
	}
	byToken := lo.SliceToMap(accessTokenIDs, toEmptyConnections)
	if len(accessTokenIDs) == 0 {
		return byToken, nil
	}
	rows, err := s.q.SimpleFinConnections(ctx, dbgen.SimpleFinConnectionsParams{AccessTokenIds: accessTokenIDs})
	if err != nil {
		return nil, err
	}
	if len(rows) == 0 {
		return byToken, nil
	}
	toConnID := func(row dbgen.SimpleFinConnectionsRow) int64 { return row.ID }
	accountsByConnID, err := s.accountsBySimpleFinConnectionIDs(ctx, u.Map(rows, toConnID))
	if err != nil {
		return nil, fmt.Errorf("load simplefin connection accounts: %w", err)
	}
	for _, row := range rows {
		conn := simpleFinConnectionFromSQLRow(row, accountsByConnID)
		byToken[row.AccessTokenID] = append(byToken[row.AccessTokenID], conn)
	}
	return byToken, nil
}

func (s *Store) SimpleFinConnectionByConnID(ctx context.Context, connID int64) (*model.SimpleFinConnection, error) {
	connections, err := s.SimpleFinConnectionsByConnIDs(ctx, []int64{connID})
	return dbutil.MapSingle(connections, err, connID)
}

func (s *Store) SimpleFinConnectionIDsByExternalID(
	ctx context.Context,
	tokenID int64,
	externalIDs []string,
) (map[string]int64, error) {
	if len(externalIDs) == 0 {
		return map[string]int64{}, nil
	}
	rows, err := s.q.SimpleFinConnections(ctx, dbgen.SimpleFinConnectionsParams{
		AccessTokenIds: []int64{tokenID},
		ExternalIds:    externalIDs,
	})
	if err != nil {
		return nil, err
	}
	toIDByExternalID := func(row dbgen.SimpleFinConnectionsRow) (string, int64) {
		return row.ExternalID, row.ID
	}
	return lo.Associate(rows, toIDByExternalID), nil
}

func (s *Store) SimpleFinConnectionsByConnIDs(ctx context.Context, connIDs []int64) (map[int64]*model.SimpleFinConnection, error) {
	if len(connIDs) == 0 {
		return map[int64]*model.SimpleFinConnection{}, nil
	}
	rows, err := s.q.SimpleFinConnections(ctx, dbgen.SimpleFinConnectionsParams{Ids: connIDs})
	if err != nil {
		return nil, err
	}
	accountsByConnID, err := s.accountsBySimpleFinConnectionIDs(ctx, connIDs)
	if err != nil {
		return nil, fmt.Errorf("load simplefin connection accounts: %w", err)
	}
	toConnectionByID := func(row dbgen.SimpleFinConnectionsRow) (int64, *model.SimpleFinConnection) {
		conn := simpleFinConnectionFromSQLRow(row, accountsByConnID)
		return conn.ID.Int64(), conn
	}
	return lo.Associate(rows, toConnectionByID), nil
}

func (s *Store) accountsBySimpleFinConnectionIDs(ctx context.Context, connIDs []int64) (map[int64][]*model.Account, error) {
	return s.accountsByConnectionSource(ctx, accounts.SourceTableSimpleFinConnection, connIDs)
}

func simpleFinConnectionFromSQLRow(row dbgen.SimpleFinConnectionsRow, accountsByConnID map[int64][]*model.Account) *model.SimpleFinConnection {
	return &model.SimpleFinConnection{
		ID:           model.New(model.GlobalIDSimpleFinConnection, row.ID),
		OrgDomain:    row.OrgDomain,
		OrgURL:       row.OrgUrl,
		Accounts:     lo.Ternary(accountsByConnID[row.ID] != nil, accountsByConnID[row.ID], []*model.Account{}),
		LastSyncedAt: row.LastSyncedAt,
		CreatedAt:    row.CreatedAt,
		UpdatedAt:    row.UpdatedAt,
	}
}

func (s *Store) SetSimpleFinConnectionHealth(ctx context.Context, connID int64, state string, errMsg *string, lastSyncedAt time.Time) error {
	lastSyncedAtUTC := lastSyncedAt.UTC()
	return s.q.SetSimpleFinConnectionHealth(ctx, dbgen.SetSimpleFinConnectionHealthParams{
		ID:                 connID,
		HealthState:        state,
		HealthErrorMessage: errMsg,
		LastSyncedAt:       &lastSyncedAtUTC,
	})
}

File: server-rs/src/accounts/store/simplefin.rs
use std::collections::HashMap;

use anyhow::{Context, Result};
use chrono::{DateTime, Utc};
use sqlx::{Executor, Sqlite, SqlitePool};

use crate::{
    accounts::simplefin_types::UpsertSimpleFinConnectionParams,
    accounts::{SimpleFinConnectionRecord, SourceTable},
    database::{self, queries},
};

use super::accounts_by_connection_sources;

pub async fn link_simple_fin_connection(
    pool: &SqlitePool,
    connection: &UpsertSimpleFinConnectionParams,
) -> Result<(i64, i64)> {
    let connection = connection.clone();
    database::with_tx(pool, |transaction| {
        Box::pin(async move {
            let simple_fin_connection_id = queries::upsert_simple_fin_connection(
                &mut **transaction,
                queries::UpsertSimpleFinConnectionParams {
                    external_id: &connection.external_id,
                    access_token_id: connection.access_token_id,
                    org_id: connection.org_id.as_deref(),
                    org_domain: connection.org_domain.as_deref(),
                    org_url: connection.org_url.as_deref(),
                    sfin_url: connection.sfin_url.as_deref(),
                    logo_url: connection.logo_url.as_deref(),
                },
            )
            .await?
            .id;
            let source_table = SourceTable::SimpleFinConnections.to_string();
            let connection_id = queries::upsert_connection(
                &mut **transaction,
                queries::UpsertConnectionParams {
                    source_table: &source_table,
                    source_id: simple_fin_connection_id,
                    name: (!connection.name.is_empty()).then_some(connection.name.as_str()),
                    owner_id: connection.owner_id,
                },
            )
            .await
            .context("upsert simplefin connection row")?
            .id;
            Ok((simple_fin_connection_id, connection_id))
        })
    })
    .await
}

pub async fn simple_fin_connections_by_token_ids(
    pool: &SqlitePool,
    access_token_ids: &[i64],
) -> Result<HashMap<i64, Vec<SimpleFinConnectionRecord>>> {
    let mut by_token = access_token_ids
        .iter()
        .copied()
        .map(|id| (id, Vec::new()))
        .collect::<HashMap<_, _>>();
    if access_token_ids.is_empty() {
        return Ok(by_token);
    }
    let rows = queries::simple_fin_connections(
        pool,
        queries::SimpleFinConnectionsParams {
            access_token_ids: Some(access_token_ids),
            ..Default::default()
        },
    )
    .await?;
    if rows.is_empty() {
        return Ok(by_token);
    }
    let connection_ids = rows.iter().map(|row| row.id).collect::<Vec<_>>();
    let accounts = accounts_by_connection_sources(pool, SourceTable::SimpleFinConnections, &connection_ids)
        .await
        .context("load simplefin connection accounts")?;
    for row in rows {
        let access_token_id = row.access_token_id;
        let connection_accounts = accounts.get(&row.id).cloned().unwrap_or_default();
        by_token
            .entry(access_token_id)
            .or_default()
            .push((row, connection_accounts).into());
    }
    Ok(by_token)
}

pub async fn simple_fin_connection_by_conn_id(
    pool: &SqlitePool,
    conn_id: i64,
) -> Result<Option<SimpleFinConnectionRecord>> {
    let connections = simple_fin_connections_by_conn_ids(pool, &[conn_id]).await?;
    Ok(connections.into_values().next())
}

pub async fn simple_fin_connection_ids_by_external_id(
    executor: impl Executor<'_, Database = Sqlite>,
    token_id: i64,
    external_ids: &[String],
) -> Result<HashMap<String, i64>> {
    if external_ids.is_empty() {
        return Ok(HashMap::new());
    }
    let token_ids = [token_id];
    queries::simple_fin_connections(
        executor,
        queries::SimpleFinConnectionsParams {
            access_token_ids: Some(&token_ids),
            external_ids: Some(external_ids),
            ..Default::default()
        },
    )
    .await
    .map(|rows| rows.into_iter().map(|row| (row.external_id, row.id)).collect())
    .map_err(Into::into)
}

pub async fn simple_fin_connections_by_conn_ids(
    pool: &SqlitePool,
    conn_ids: &[i64],
) -> Result<HashMap<i64, SimpleFinConnectionRecord>> {
    if conn_ids.is_empty() {
        return Ok(HashMap::new());
    }
    let rows = queries::simple_fin_connections(
        pool,
        queries::SimpleFinConnectionsParams {
            ids: Some(conn_ids),
            ..Default::default()
        },
    )
    .await?;
    let accounts = accounts_by_connection_sources(pool, SourceTable::SimpleFinConnections, conn_ids)
        .await
        .context("load simplefin connection accounts")?;
    Ok(rows
        .into_iter()
        .map(|row| {
            let id = row.id;
            let connection_accounts = accounts.get(&id).cloned().unwrap_or_default();
            (id, (row, connection_accounts).into())
        })
        .collect())
}

pub async fn set_simple_fin_connection_health(
    executor: impl Executor<'_, Database = Sqlite>,
    conn_id: i64,
    state: &str,
    error_message: Option<&str>,
    last_synced_at: DateTime<Utc>,
) -> Result<()> {
    queries::set_simple_fin_connection_health(
        executor,
        queries::SetSimpleFinConnectionHealthParams {
            health_state: state,
            health_error_message: error_message,
            last_synced_at: Some(last_synced_at.into()),
            id: conn_id,
        },
    )
    .await
    .map_err(Into::into)
}

#[cfg(test)]
mod tests {
    use anyhow::Result;

    use crate::database::dbtest;

    use super::*;

    #[tokio::test]
    async fn empty_batch_lookups_return_empty_maps() -> Result<()> {
        let pool = dbtest::open().await?;
        assert!(simple_fin_connections_by_token_ids(&pool, &[]).await?.is_empty());
        assert!(simple_fin_connections_by_conn_ids(&pool, &[]).await?.is_empty());
        assert!(
            simple_fin_connection_ids_by_external_id(&pool, 1, &[])
                .await?
                .is_empty()
        );
        Ok(())
    }
}
